src/exif.c

```diff
@@ -0,0 +1,133 @@
+    }
+
+    if (*exifTiffHeaderOffset >= exif->size - 4) {
+        // Couldn't find the TIFF header
+        return AVIF_RESULT_INVALID_EXIF_PAYLOAD;
+    }
+    return AVIF_RESULT_OK;
+}
+
+avifResult avifExtractExifOrientation(const avifRWData * exif, avifTransformFlags * flags, avifImageRotation * irot, avifImageMirror * imir)
+{
+    uint32_t exifTiffHeaderOffset;
+    const avifResult result = avifExtractExifTiffHeaderOffset(exif, &exifTiffHeaderOffset);
+    if (result != AVIF_RESULT_OK) {
+        // Couldn't find the TIFF header
+        return result;
+    }
+
+    avifROData raw = { exif->data + exifTiffHeaderOffset, exif->size - exifTiffHeaderOffset };
+    const avifBool isLittleEndian = (raw.data[0] == 'I');
+    avifROStream stream;
+    avifROStreamStart(&stream, &raw, NULL, NULL);
+
+    // TIFF Header
+    uint32_t offsetToNextIfd;
+    if (!avifROStreamSkip(&stream, 2 + 2) ||                                         // Byte Order ("II" or "MM" then "42")
+        !avifROStreamReadU32Endianness(&stream, &offsetToNextIfd, isLittleEndian) || // Offset to 0th IFD
+        offsetToNextIfd < 2 + 2 + 4) {
+        return AVIF_RESULT_INVALID_EXIF_PAYLOAD;
+    }
+
+    while (offsetToNextIfd) { // for each IFD
+        avifROStreamSetOffset(&stream, offsetToNextIfd);
+        uint16_t fieldCount;
+        if (!avifROStreamReadU16Endianness(&stream, &fieldCount, isLittleEndian)) {
+            return AVIF_RESULT_INVALID_EXIF_PAYLOAD;
+        }
+        for (uint16_t field = 0; field < fieldCount; ++fieldCount) { // for each interoperability array
+            uint16_t tag;
+            uint16_t type;
+            uint32_t count;
+            uint16_t firstHalfOfValueOffset;
+            uint16_t secondHalfOfValueOffset;
+            if (!avifROStreamReadU16Endianness(&stream, &tag, isLittleEndian) ||
+                !avifROStreamReadU16Endianness(&stream, &type, isLittleEndian) ||
+                !avifROStreamReadU32Endianness(&stream, &count, isLittleEndian) ||
+                !avifROStreamReadU16Endianness(&stream, &firstHalfOfValueOffset, isLittleEndian) ||
+                !avifROStreamReadU16Endianness(&stream, &secondHalfOfValueOffset, isLittleEndian)) {
+                return AVIF_RESULT_INVALID_EXIF_PAYLOAD;
+            }
+            // Orientation attribute according to JEITA CP-3451C section 4.6.4 (TIFF Rev. 6.0 Attribute Information):
+            if (tag == 0x0112 && type == /*SHORT=*/0x03 && count == 0x01) { // Exif IFD
+                // Mapping from Exif orientation as defined in JEITA CP-3451C section 4.6.4.A Orientation
+                // to irot and imir boxes as defined in HEIF ISO/IEC 28002-12:2021 sections 6.5.10 and 6.5.12.
+                switch (firstHalfOfValueOffset) {
+                    case 1: // The 0th row is at the visual top of the image, and the 0th column is the visual left-hand side.
+                        *flags = AVIF_TRANSFORM_NONE;
+                        irot->angle = 0; // ignored
```

Comment:
I don't know what these "ignored" comments mean...

UPDATE: I understand it now. It means the caller will ignore the value because the corresponding bit in `*flags` is not set.

I think it may be better to just pass `avifImage *image` to this functon and let this function reads `image->exit` and modifies `image->transformFlags`, `image->irot`, and `image->imir` directly.

If you do that, then this function should be named `avifImageExtractExifOrientation` to follow libavif's naming convention for object-oriented, method-like functions.

Reply:
Another idea is to keep the current function prototype but use `flags`, `irot`, and `imir` as in/out parameters. Then case 1, case 2, etc. of this switch statement would be written as:

```
                    case 1: // The 0th row is at the visual top of the image, ...
                        return AVIF_RESULT_OK;
                    case 2: // The 0th row is at the visual top of the image, ...
                        *flags |= AVIF_TRANSFORM_IMIR;
                        imir->mode = 1;
                        return AVIF_RESULT_OK;
                    case 3: // The 0th row is at the visual bottom of the image, ...
                        *flags |= AVIF_TRANSFORM_IROT;
                        irot->angle = 2;
                        return AVIF_RESULT_OK;
                    ...
```

And we would call this function as follows:

```
            // Import Exif orientation only if not already explicitly specified by the user.
            if (!(encoder->data->imageMetadata->transformFlags & (AVIF_TRANSFORM_IROT | AVIF_TRANSFORM_IMIR))) {
                // Ignore any Exif parsing failure.
                (void)avifExtractExifOrientation(&encoder->data->imageMetadata->exif,
                                                 &encoder->data->imageMetadata->transformFlags,
                                                 &encoder->data->imageMetadata->irot,
                                                 &encoder->data->imageMetadata->imir);
            }
```

Reply:
> Another idea is to keep the current function prototype but use `flags`, `irot`, and `imir` as in/out parameters.

I did that at first, then thought that it was difficult to grasp what was the input and what was set. Also it does not make sense to set the angle without clearing the rotation, unless it is known that this function is only called when there is no `AVIF_TRANSFORM_IROT` and no `AVIF_TRANSFORM_IMIR` prior to calling it.

> I think it may be better to just pass `avifImage *image` to this functon and let this function reads `image->exit` and modifies `image->transformFlags`, `image->irot`, and `image->imir` directly.

I think the logic at the call site is slightly more obscure that way but done.

Reply:
I didn't have time to review this pull review again today. I will review it tomorrow (Wednesday).

One thing I noticed last Friday was that this pull request always imports Exif orientation. It may be better to let the user control whether Exif orientation is imported. One reason is that it will allow the encoded output to match the input `avifImage` faithfully. In that design, we will need a function that the user can call to import Exif orientation, and that function can take `avifImage *image` as input.

Reply:
> I didn't have time to review this pull review again today.

We can save time by agreeing on the general design before reviewing the implementation details.

> One thing I noticed last Friday was that this pull request always imports Exif orientation.

The default behavior should be to encode Exif orientation as irot/imir whenever possible. That is why I put directly within `avifEncoderAddImageInternal()`. It was possible to skip it but in a shady way: set `irot.angle` to `0` and use `AVIF_TRANSFORM_IROT`. It would keep the default unoriented behavior but would add an unnecessary `irot` box to the encoded bitstream.

> It may be better to let the user control whether Exif orientation is imported.

I agree. I added an `extractExifOrientationToIrotImir` argument to `avifImageSetMetadataExif()` instead of doing it in `avifEncoderAddImageInternal()`. If `avifImageSetMetadataExif()` is the only recommended way to deal with Exif metadata, the user must set that option. The downsides are: the user has to care about the irot/imir topic (no default), `avifImageSetMetadataExif(image, nullptr, 0, AVIF_TRUE)` will not clear irot/imir (easily overlooked).

The alternative would be going back to the `avifEncoderAddImageInternal()` but with an option such as `avifEncoder.extractExifOrientationToIrotImir` set to `AVIF_TRUE` by default. This alternative does not respect your constraint about matching input `avifImage` faithfully for default settings.

What do you think?

---

Somewhat related: I found the following in MIAF ISO 23000-22 section 7.3.10.

> Metadata of type Exif as specified by JEITA CP-3451 may be present in the file and shall be output by the MIAF reader; this means that a reader shall be able to transfer this data to the application. There is no need for a reader to understand the details of the metadata. **There should be no image transformations expressed by Exif (rotation, mirroring, etc.) indicated in the Exif metadata, in files encoded according to this document.** If present, they shall be ignored by a MIAF renderer, i.e. not cause a change in the presentation of associated images by a MIAF player.

Should we strip orientation data from the encoded Exif payload in libavif?

Reply:
I will comment on how to control extracting Exif orientation later.

If it is safe (i.e., not prone to software security issues) to strip orientation data from Exif, we can do that. I think keeping irot and imir in sync with the Exif orientation data is the next best thing and is good enough.

Reply:
Just a small subtle fact to think about: the Exif Orientation tag has a default value of 1 (first pixel supplied is top left) if not present, so you cannot really "strip" it. Rather than going through the acrobatics with reallocating the payload and copying chunks around, you might want to simply ensure it is set to 1 instead (i.e. instruct readers to do nothing further)?

Reply:
> I think keeping irot and imir in sync with the Exif orientation data is the next best thing and is good enough.

Let's do only that in this PR then, I agree. It was just a thought.

> the Exif Orientation tag has a default value of 1

Indeed, it is defined in JEITA CP-3451C section 4.6.4.A Orientation.